### DISCUSSION TITLE
Support for recursive records

### DISCUSSION THREAD

**Discussion by suimong** [nushell/nushell#11255] [Ideas]
Just would like to share my two cents on a feature that IMHO would enhance the experience of using nushell as a programming language even more: recursive records, as is available in Nix, Nickel etc.

Ideally, we should able to write nushell like the following (which is syntactically valid but does not work as intended ATM because nushell will try to find `$dirs`, `$root` in the outer scope):

```nushell
let ops = {
  dirs: {
    root: "./ops",
    configs: ([$root "configs"] | path join),  
  },
  files: {
    custom_bashrc: ([$dirs.configs "custom.bashrc"] | path join)
  },
}
```

Whereas in Nix, one could write:

```nix
let 
  ops = rec {
    dirs = rec {
      root = "./ops";
      configs = "${root}/configs";
    };
    files = {
      custom_bashrc = "${dirs.configs}/custom.bashrc";
    };
  };
in
ops
```

To achieve the same in the current nushell implementation, one has to split the records into multiple let bindings and order them based on dependencies. 

IMHO, the non-recursive records has a few ergonomic downsides:
1. the variable names are now in the outer scope, with a much bigger chance of name clash for common words like "root" "configs" etc., therefore one has to resort to environment-variable-style dash-concatenated variable name, e.g. `ops_dirs_configs`, `ops_files_custom_bashrc`, which kind of diminish the utility of nushell's capability of having typed data.
2. user's ability to organize the structures of records based on the business logic (as is the case with the above example) is impaired, because the order of multiple let bindings are dictated by the dependencies between variables, which should not matter for pure evaluation.

Love to know the thoughts from the community. cheers.

**Comment by fdncred** [comment]
Thanks for the interesting idea. I'm no parser expert, so I could be wrong, but I'd guess this would not happen because for `$dirs` and `$root` to be visible, you'd have to evaluate each line item as it is written. This goes against these points on how nushell works https://www.nushell.sh/book/thinking_in_nu.html and https://www.nushell.sh/book/how_nushell_code_gets_run.html.

**Reply by suimong** [reply]
Hi @fdncred , thanks for the quick reply.

I read the pages you referenced, and it is not clear to me how exactly the proposed idea would violate the way nushell runs the code.

I know little about nushell's internals, and not a parser expert either, but I do feel that certain levels of "lazy evaluation" is possible, e.g. with Nix and Nickel. Both are pure functional language, with no "eval" function, and can perform lazy evaluation. I gathered a small [repo](https://github.com/suimong/nickel-lazy-demo) for demonstration if you are interested.

**Comment by sophiajt** [comment]
Thanks for jumping in and exploring the design space.

Having data structures with recursion is outside of the scope of Nushell 1.0, which is our current focus. We may explore it later when we return to exploring how the data model works. For now, though, we're going to ship the model we have, fixing bugs where necessary.

**Reply by suimong** [reply]
Hi @sophiajt, I'm definitely in favor of having the 1.0 release! The idea of having records with recursion only occurs to me when I'm working on a nix configuration where a bash script is embedded inside the nix file, which by merely using string interpolation, benefitting from nix's recursive record feature, and I thought why not using nushell for the scripting. Writing this thread got me thinking about the problem a bit more, and I realized that the main source of data in nushell world comes from somewhere else, which by nature is "static" and "recursion" is not a thing; this separates clearly from Nix/Nickel's main use cases i.e. configuration management, which by nature requires manual encoding of data and logic and is where recursion shines. 

That being said, combining the power of two worlds might spark much more interesting use cases. For example, one could build a nushell plugin that implements a "from nickel" command, then using nu's piping and query abilities to lazily query the nickel configuration file. Further brainstorming on the idea, when scripting, nu could implement a special "import"/"use" for nickel file, while the lsp could parse the nickel file for auto-completion (this might even be "readily available" by relying on nickel's lsp itself).

Hope you get the chance to revisit this topic in the future!
